lib/miq_expression.rb

```diff
@@ -1427,23 +1425,93 @@ def to_arel(exp, tz)
     end
   end
 
-  def extract_where_values(klass, scope)
-    relation = ActiveRecord::Relation.new(klass, klass.arel_table, klass.predicate_builder)
-    relation = relation.instance_eval(&scope)
+  def subquery_for_contains(field, subquery_limiter)
+    # base_class removes the default scopes. They are already in the main query and not needed in this subquery
+    main_model = field.model.base_class
 
-    begin
-      # This is basically ActiveRecord::Relation#to_sql, only using our
-      # custom visitor instance
-
-      connection = klass.connection
-      visitor    = WhereExtractionVisitor.new(connection)
-
-      arel  = relation.arel
-      binds = relation.bound_attributes
-      binds = binds.collect(&:value_for_database)
-      binds.map! { |value| connection.quote(value) }
-      collect = visitor.accept(arel.ast, Arel::Collectors::Bind.new)
-      collect.substitute_binds(binds).join
+    to_ref = field.reflections.first
+    join_keys = to_ref.join_keys
+    primary_attribute = field.model.arel_attribute(join_keys.foreign_key)
+    secondary_attribute = to_ref.klass.arel_attribute(join_keys.key)
+
+    # algorithm:
+    # 1. generate subquery as part of our main table since it has all the concepts we need.
+    # 2. remove main_table from this sub query.
+    # 3. move the secondary_table from the JOIN to the FROM clause
+    # 4. add miq_expression filter to sub query.
+    # 5. if the join on clause has an AND, add the second part back to the query
+    # 6. replace SELECT 1 place holder with secondary_table id
+
+    # 1. SELECT 1 FROM main_table JOIN secondary_table ON main_table.id = secondary_table.main_id AND type = 'type'
+    includes_associations = field.reflections.inject({}) { |i, k| {k.name => i} }
+    query_relation = main_model.joins(includes_associations).select(1).distinct
+    query = query_relation.arel
+
+    # 2. SELECT 1 FROM main_table [ ]
+    join = query.source.right.shift
+    # 3. SELECT 1 FROM [secondary_table]
+    query.source.left = join.left
+    # 4. SELECT 1 FROM secondary_table [WHERE name = 'foo']
+    query.where(subquery_limiter)
+
+    # 5. SELECT 1 FROM secondary_table WHERE name = 'foo' [AND type = 'type']
+    join_on_clause = join.right.expr
+
+    sub_query = extract_key(join_on_clause, primary_attribute, secondary_attribute)
+    sub_query = remove_binds_from_query(main_model.connection, sub_query, query_relation.bound_attributes)
+    query.where(sub_query) if sub_query
+
+    # 6. SELECT [secondary_table_id] FROM secondary_table WHERE name = 'foo' AND type = 'type'
+    query_relation.projections.shift
+    query_relation.project(secondary_attribute)
+
+    primary_attribute.in(query)
+  end
+
+  def extract_key(query, field1, field2)
```

Comment:
I wrote a bunch of comments down in the specs, but I found myself having a hard time understanding what `.extract_key` was doing, and what it's return value was supposed to be.

From what I can tell, it is used to extract the "ON table_1.id = table_2.table_1_id" portion of the query, where ever it may be, but it is far from obvious that it is doing that.  Furthermore, it isn't obvious that `nil` means the query "only contains the `table_1.id = table_2.table_1_id` portion".

Some comments at a minimum would be nice, and adding more some comments in the specs to define what type of query is being test would also be welcome.

Reply:
renamed this `remove_equality` and added comments

Reply:
looking into converting back into a visitor